File: gto-support-db/src/testCommon/java/org/ccci/gto/android/common/db/TestDatabase.java
package org.ccci.gto.android.common.db;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import androidx.annotation.NonNull;

import org.ccci.gto.android.common.db.CommonTables.LastSyncTable;
import org.ccci.gto.android.common.db.Contract.CompoundTable;
import org.ccci.gto.android.common.db.Contract.RootTable;

class TestDatabase extends WalSQLiteOpenHelper {
    private TestDatabase(@NonNull final Context context) {
        super(context, "test_db", null, 1);
        resetDatabase(getWritableDatabase());
    }

    private static TestDatabase sInstance;
    static TestDatabase getInstance(@NonNull final Context context) {
        synchronized (TestDatabase.class) {
            if (sInstance == null) {
                sInstance = new TestDatabase(context.getApplicationContext());
            }
            return sInstance;
        }
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(RootTable.SQL_CREATE_TABLE);
        db.execSQL(CompoundTable.SQL_CREATE_TABLE);
        db.execSQL(LastSyncTable.SQL_CREATE_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        throw new IllegalStateException("onUpgrade should not be triggered");
    }

    private void resetDatabase(final SQLiteDatabase db) {
        try {
            db.beginTransaction();

            db.execSQL(RootTable.SQL_DELETE_TABLE);
            db.execSQL(CompoundTable.SQL_DELETE_TABLE);
            db.execSQL(LastSyncTable.SQL_DELETE_TABLE);

            onCreate(db);

            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }
    }
}

File: gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/TestDatabase.kt
package org.ccci.gto.android.common.db

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import org.ccci.gto.android.common.db.CommonTables.LastSyncTable
import org.ccci.gto.android.common.db.Contract.CompoundTable
import org.ccci.gto.android.common.db.Contract.RootTable

internal class TestDatabase private constructor(context: Context) : WalSQLiteOpenHelper(context, "test_db", null, 1) {
    init {
        resetDatabase(writableDatabase)
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.apply {
            execSQL(RootTable.SQL_CREATE_TABLE)
            execSQL(CompoundTable.SQL_CREATE_TABLE)
            execSQL(LastSyncTable.SQL_CREATE_TABLE)
        }
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) =
        error("onUpgrade should not be triggered")

    private fun resetDatabase(db: SQLiteDatabase) {
        try {
            db.beginTransaction()
            db.execSQL(RootTable.SQL_DELETE_TABLE)
            db.execSQL(CompoundTable.SQL_DELETE_TABLE)
            db.execSQL(LastSyncTable.SQL_DELETE_TABLE)
            onCreate(db)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    companion object {
        private var instance: TestDatabase? = null

        @JvmStatic
        @Synchronized
        fun getInstance(context: Context) = instance ?: TestDatabase(context.applicationContext).also { instance = it }
    }
}
